cmd/root/exercise.go

```diff
@@ -0,0 +1,301 @@
+	typedLast := rune(m.typedText[l-1])
+
+	if exLast != typedLast {
+		return false
+	}
+	return true
+}
+
+// Converts the exercise model to a Rep, in preparation for
+// inserting it into the database.
+func exerciseModelToRep(m exerciseModel) db.Rep {
+	return db.Rep{
+		Hash:   util.MD5Hash(m.exercise.text),
+		Start:  m.events[0].Ts,
+		End:    m.events[len(m.events)-1].Ts,
+		Name:   m.exercise.name,
+		Lang:   util.Lang(m.exercise.name),
+		Wpm:    wpm(m.events),
+		Raw:    wpmRaw(m.events),
+		Dur:    duration(m.events),
+		Acc:    accuracy(m.events),
+		Miss:   numMistakes(m.events),
+		Errs:   numUncorrectedErrors(m.events),
+		Events: m.events,
+	}
+}
+func (m exerciseModel) Init() tea.Cmd {
+	return nil
+}
+
+func (m exerciseModel) Update(msg tea.Msg) (tea.Model, tea.Cmd) {
+	switch msg := msg.(type) {
+	case tea.KeyMsg:
+		var currTyped string
+		currI := len(m.typedText)
+		currExpected := event.RuneToEventExpected(rune(m.exercise.text[currI]))
+		switch msg.Type {
+		case tea.KeyCtrlC:
+			m.quitEarly = true
+			return m, tea.Quit
+		case tea.KeyBackspace:
+			currTyped = event.TeaKeyMsgToEventTyped(msg)
+			m = m.deleteRuneFromTypedText()
+			// Create delete event and add it to events
+			m.events = append(m.events, event.NewEvent("backspace", "", currI))
+		case tea.KeyRunes, tea.KeySpace, tea.KeyEnter:
+			currTyped = event.TeaKeyMsgToEventTyped(msg)
+
+			if m.startTime.IsZero() {
+				m.startTime = time.Now()
+			}
+			if msg.Type == tea.KeyEnter {
+				m = m.addRuneToTypedText(c.Enter)
+			} else {
+				m = m.addRuneToTypedText(msg.Runes[0])
+			}
+			m.events = append(m.events, event.NewEvent(currTyped, currExpected, currI))
+			if m.finished() {
+				m.endTime = time.Now()
+				return m, tea.Quit
+			}
+		}
+	}
+	return m, nil
+}
+
+// Displays the text for the typing exercise.
+// Hides the view once the exercise is complete or the user quits early.
+func (m exerciseModel) View() (s string) {
+	if !m.finished() {
+		s += "\n"
+		s += m.exerciseNameView()
+		s += "\n\n"
+		s += m.exerciseTextView()
+		s += "\n"
+
+		currKeyI := len(m.typedText)
+		currKey := m.exercise.text[currKeyI]
+		s += qwerty.render(string(currKey))
+		s += "\n"
+		s += fingerView(qwerty.fingersMargin, '*', rune(currKey))
+	}
+	return
+}
+
+// Runs the exercise, which does the following
+//
+// 1. Runs the bubbletea interactive typing application.
+//
+// 2. Depending on the outcome of the exercise, either completes
+// or returns an error.
+//
+// 3. If the exercise is completed, gather the results, print them, and
+// save them to the database
+func run(exercise exercise) {
+	teaModel, err := tea.NewProgram(exerciseModel{
+		exercise:  exercise,
+		typedText: "",
+		quitEarly: false,
+		startTime: time.Time{},
+		endTime:   time.Time{},
+		events:    []event.Event{},
+	}).Run()
+	if err != nil {
+		fmt.Printf("Error running typing exercise: %v\n", err)
+		os.Exit(1)
+	}
+
+	exModel, ok := teaModel.(exerciseModel)
```

Comment:
Do I really need to make this cast? can't I just take the teaModel and do something like exerciseModel(teaModel) when passing it into other functions?

Reply:
Needs type assertion, cannot just implicitly convert it.